Sources/HiPMobile/DataAccessLayer/IJoinEntity.cs

```diff
@@ -0,0 +1,170 @@
+
+    /// <summary>
+    /// Supports relational many-to-many relationships of the same entity type. Such a
+    /// many-to-many relationship is usually modeled through a "join type/table" J containing
+    /// one row for each connection. This type J should implement this interface.
+    /// </summary>
+    public interface IJoinEntitySameType<TEntity>
+    {
+        TEntity this[JoinSide side] { get; set; }
+    }
+
+    public enum JoinSide { A, B }
+
+    /// <summary>
+    /// Supports relational many-to-many relationships of the same entity type. Such a
+    /// many-to-many relationship is usually modeled through a "join type/table" J
+    /// containing one row for each connection. This facade is a "shortcut" which allows
+    /// to get, establish and remove connections without the need to deal with the join type J.
+    /// </summary>
+    /// <typeparam name="TEntity"></typeparam>
+    /// <typeparam name="TJoinEntity">The join type</typeparam>
+    public struct JoinCollectionFacade<TEntity, TJoinEntity> : ICollection<TEntity>
+        where TJoinEntity : IJoinEntitySameType<TEntity>, new()
+    {
+        private readonly TEntity ownerEntity;
+        private readonly ICollection<TJoinEntity> collection;
+        private readonly JoinSide navigationTarget;
+        private static int target;
+
+        public JoinCollectionFacade(TEntity ownerEntity, ICollection<TJoinEntity> collection, JoinSide navigationTarget)
+        {
+            this.ownerEntity = ownerEntity;
+            this.collection = collection;
+            this.navigationTarget = navigationTarget;
+        }
+
+        public IEnumerator<TEntity> GetEnumerator()
+        {
+            var target = navigationTarget;
+            return collection.Select(e => e[target]).GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public void Add(TEntity item)
+        {
+            var entity = new TJoinEntity();
+            entity[navigationTarget] = item;
+            entity[navigationTarget == JoinSide.A ? JoinSide.B : JoinSide.A] = ownerEntity;
+            collection.Add(entity);
+        }
+
+        public void Clear() => collection.Clear();
+
+        public bool Contains(TEntity item)
+        {
+            var target = navigationTarget;
+            return collection.Any(e => Equals(item, e, target));
+        }
+
+        public void CopyTo(TEntity[] array, int arrayIndex)
+        {
+            if (array == null)
+                throw new ArgumentNullException(nameof(array));
+
+            if (arrayIndex < 0)
+                throw new ArgumentOutOfRangeException(nameof(arrayIndex));
+
+            if (collection.Count > array.Length - arrayIndex)
+                throw new ArgumentException("Insufficient space in target array");
+
+            var i = 0;
+            var target = navigationTarget;
+            foreach (var item in collection)
+                array[arrayIndex + i++] = item[target];
```

Comment:
Using `i++` inside the indexing operator can be confusing (especially for people who don't code every day), please move it into a seperate line to also make order of operations clear.